source/common/modules/markdown-editor/index.ts

```diff
@@ -710,11 +704,12 @@ export default class MarkdownEditor extends EventEmitter {
           const anchorLine = this._instance.state.doc.lineAt(sel.anchor)
           const headLine = this._instance.state.doc.lineAt(sel.head)
           const selContent = this._instance.state.sliceDoc(sel.from, sel.to)
+          const ast = markdownToAST(selContent)
```

Comment:
It is relatively costly to parse the markdown content, so as an optimization you may want to move the words/char count calculation to a queue with a very short delay (if I start selecting everything in a document by using Ctrl + Line down, then I can easily get > 25% CPU on a relatively beefy machine).

Reply:
Very good catch! However, since asynchronous callbacks in this case are always tricky to do and would increase the complexity again, I think I have had an idea:

In a previous commit, I have added the ability of `markdownToAST` to use a pre-parsed SyntaxTree, if you pass one as the second parameter. This completely obliterates the need for the AST parser to create a parse tree by itself. Since CodeMirror always keeps at least a partial parse tree (which will include at least the selections), what I have done now is to pass said partial tree to the function, which cut down the parse time by a factor of 5.

A second optimization is that I have now exchanged the "countChars" and "countWords" function with a novel function, "countAll", which has the char and word counter share the computationally heavy preprocessing (and, prospectively, will replace the individual functions), thus further cutting down the remaining time in half.

I have run a few performance test: The unoptimized code took about ~60ms per each keydown event (i.e. each move to the next line), the optimized code now cuts this down to ~12ms, of which the actual counting of the characters and words makes up approximately half to 1ms.

Do you think this suffices, or do you still think it may negatively impact users' computers?

Please feel free to have a look at the newly optimized code.

If you think that this still puts too much load onto computers (thinking especially of potato PCs, on which Zettlr should run smooth as well), I can reduce the update interval with a debouncer.

Reply:
I think this should suffice. It's a non-blocking operation anyway, so it should be fine.

Reply:
Alright, thank you very much for your thoughts on this! In this case, I'll merge the PR now!